src/layer_env.rs

```diff
@@ -0,0 +1,528 @@
+                        previous_value.push(self.delimiter_for(&entry.name));
+                    }
+
+                    previous_value.push(&entry.value);
+
+                    result_env.insert(&entry.name, previous_value);
+                }
+                ModificationBehavior::Prepend => {
+                    let previous_value = result_env.get(&entry.name).unwrap_or(OsString::new());
+
+                    let mut new_value = OsString::new();
+                    new_value.push(&entry.value);
+
+                    if !previous_value.is_empty() {
+                        new_value.push(self.delimiter_for(&entry.name));
+                        new_value.push(previous_value);
+                    }
+
+                    result_env.insert(&entry.name, new_value);
+                }
+                _ => (),
+            };
+        }
+
+        result_env
+    }
+
+    fn delimiter_for(&self, key: impl AsRef<OsStr>) -> OsString {
+        self.entries
+            .iter()
+            .find(|entry| {
+                entry.name == key.as_ref()
+                    && entry.modification_behavior == ModificationBehavior::Delimiter
+            })
+            .map(|entry| entry.value.clone())
+            .unwrap_or(OsString::new())
+    }
+
+    fn read_from_env_dir(path: impl AsRef<Path>) -> Result<Self, std::io::Error> {
+        let mut layer_env = Self::empty();
+
+        for dir_entry in fs::read_dir(path.as_ref())? {
+            let path = dir_entry?.path();
+
+            // Rely on the Rust standard library for splitting stem and extension. Since paths
+            // are not necessarily UTF-8 encoded, this is not as trivial as it might look like.
+            // Think twice before changing this.
+            let file_name_stem = path.file_stem();
+            let file_name_extension = path.extension();
+
+            // The CNB spec explicitly states:
+            //
+            // > File contents MUST NOT be evaluated by a shell or otherwise modified before
+            // > inclusion in environment variable values.
+            // > https://github.com/buildpacks/spec/blob/a9f64de9c78022aa7a5091077a765f932d7afe42/buildpack.md#provided-by-the-buildpacks
+            //
+            // This should include parsing the contents with an assumed charset and later emitting
+            // the raw bytes of that encoding as it might change the actual data. Since this is not
+            // explicitly written in the spec, we read through the the reference implementation and
+            // determined that it also treats the file contents as raw bytes.
+            // See: https://github.com/buildpacks/lifecycle/blob/a7428a55c2a14d8a37e84285b95dc63192e3264e/env/env.go#L73-L106
+            use std::os::unix::ffi::OsStringExt;
+            let file_contents = OsString::from_vec(fs::read(&path)?);
+
+            if let Some(file_name_stem) = file_name_stem {
+                let r#type = match file_name_extension {
```

Comment:
The use of `r#type` confused me for a moment, until I wondered whether it was due to `type` being a reserved keyword, and thus I found the raw identifier syntax here:
https://doc.rust-lang.org/reference/identifiers.html

Should we use a different variable name to avoid this confusing others too? eg `behaviour_type` ?